Add Worklist implementations

haved opened the pull request:
To avoid one big PR, here is the set of worklist implementations I have, used by the new Andersen solver.

I decided to use tightly packed vectors instead of HashMaps / HashSets. This can of course be benchmarked later.

phate reviewed (commented):
I got interrupted and could not finish my review. I will need to have another look at it. Just leaving what I have so far.

I am a little bit worried about the template parameter only being of integer as this would limit the usefulness severely.

haved commented:
I implemented your suggested changes, and fixed the tests (always test with `--enable-asserts`...).

While I think performance might be better with `std::vector`s, it does make the interface less general, so I moved it to `util::HashSet` and `std::unordered_map` for now. The worklist is a tight part of the algorithm, so it will be interesting to see in the future if it has an effect.

phate commented on jlm/llvm/opt/alias-analyses/Worklist.hpp:
```diff
@@ -0,0 +1,220 @@
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_WORKLIST_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_WORKLIST_HPP
```
Could we move the entire file to to `util`? These implementations are useful for other parts in the compiler as well.

phate reviewed (changes requested):
The main issue I have is that this should live in libutil and not in the alias-analysis subdirectory. This is quite useful for other parts in the compiler.

I agree with you that using `std::vector` would be more performant due to caching effects, but that should be possible by using template specialization:

```
template<>
LifeWorklist<size_t> : public Worklist<size_t>
{
...
}
```
(written by heart, syntax might be off)
where you provide another implementation with `std::vector`. In other words, we can have the cake and also eat it. The cost is another implementation (and tests).

haved commented on jlm/llvm/opt/alias-analyses/Worklist.hpp:
```diff
@@ -0,0 +1,220 @@
+};
+
+/**
+ * Worklist implementation using a priority queue.
+ * The next work item to fire is the one that was least recently fired (LRF).
+ * @tparam T the type of the work items.
+ * @see Worklist
+ */
+template<typename T>
+class LrfWorklist final : public Worklist<T>
```
I added some more description of it, as well as references to the original sources.

I see what you mean regarding the sources. I was only able to find [Worklist management strategies for dataflow analysis](https://scholar.google.no/scholar?q=Worklist+management+strategies+for+dataflow+analysis&hl=no&as_sdt=0&as_vis=1&oi=scholart) as a postscript file.

haved commented:
I moved it to util, and started on specialization for integer types, but ended up with SFINAE fluff that didn't look that nice by the time it worked, so I'm trying to make it a bit more sane first

haved commented on jlm/llvm/opt/alias-analyses/Worklist.hpp:
```diff
@@ -0,0 +1,220 @@
+  }
+
+  T
+  PopWorkItem() override
+  {
+    JLM_ASSERT(HasMoreWorkItems());
+    T item = WorkItems_.top();
+    WorkItems_.pop();
+    OnList_.Remove(item);
+    return item;
+  }
```
Since we are copying all the types into both the OnList and the queue/stack itself, the class really shouldn't be used with types that are not trivially copyable. In that case the classes should be written to not make copies, and use stable references.

haved commented:
I ended up not adding specializations for bounded integer typed Worklists now, as it always became quite ugly. The best solution I found was to create a class with the same interface as `util::HashSet<T>`, but enforce `T` to be integer and the bounds to be provided in the ctor. The same would have to be done for `util::HashMap<K,V>`, but we currently don't even have our own `util::HashMap<K,V>`, and it felt a bit overkill to add that now.

Instead, I will investigate if preallocated sequential vectors makes a difference once the Worklist actually is in use.
